cmake: gcc/clang: enable colour diagnostics in terminal outptut

arvinf opened the pull request:
When gcc/clang output is piped through ninja the compiler disables colour diagnostics. Using `-fdiagnostics-color` forces the compiler to enable colour output.

Colour highlighted error and warning messages are useful during development.

tejlmand commented on cmake/compiler/gcc/target_color.cmake:
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
+list(APPEND CMAKE_ASM_FLAGS -fdiagnostics-color=always)
+list(APPEND CMAKE_C_FLAGS -fdiagnostics-color=always)
+list(APPEND CMAKE_CXX_FLAGS -fdiagnostics-color=always)
```
This should as minimum use: `zephyr_cc_option` or `zephyr_compile_options`

tejlmand reviewed (changes requested):
Thanks for this.
A couple of direct comments, but also are more general comment.

This is only needed when using the Ninja generator, when using make, color output with gcc works without those flags. (Although they don't harm when using `make`)

Secondly, is it guaranteed that this flag works with all toolchains that are handled as gcc ?
(They should if they are truly based on gcc 4.9 or newer.)
https://docs.zephyrproject.org/latest/getting_started/toolchain_3rd_party_x_compilers.html#crosstool-ng
 
So far I tested with `gnuarmemb` and `Zephyr SDK`.

tejlmand commented:
In general compile flags should be controllable by user using Kconfig, but  `diagnostics-color` flag is purely color output and thus doesn't affect created object files.

@nashif Should we allow this particular flag as always-on.

tejlmand commented:
@arvinf I see you did a force-push (probably a rebase), but you seem to have forgotten your own changes.

arvinf commented:
Thank you. That was just a rebase.

I've made the changes you requested and re-pushed it.

tejlmand commented on cmake/compiler/clang/target_color.cmake:
```diff
@@ -0,0 +1,2 @@
+# SPDX-License-Identifier: Apache-2.0
+zephyr_compile_options(-fcolor-diagnostics)
```
could you add a newline at end of line ?

tejlmand commented on cmake/compiler/gcc/target_color.cmake:
```diff
@@ -0,0 +1,2 @@
+# SPDX-License-Identifier: Apache-2.0
+zephyr_compile_options(-fdiagnostics-color=always)
```
could you add a newline at end of line ?

tejlmand reviewed (commented):
Looks good.
Just add newline, so it doesn't standout.

arvinf commented:
@nashif are there plans to merge this in? Could you please review?

tejlmand commented:
@arvinf seems this got closed by mistake :disappointed: 

Could you please rebase so we can get it in :+1:

arvinf committed:
cmake: gcc/clang: enable colour diagnostics in terminal output

When compiler results are piped through a non-terminal (e.g. ninja)
the compiler disables colour diagnostics. Using `-fdiagnostics-color`
forces the compiler to enable colour output. This flag is set for
clang and gcc when `ZEPHYR_BUILD_COLOUR_DIAGNOSTIC` environment
variable is set when a clean build is started.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

arvinf commented:
Rebased and updated. This feature is only enabled when environment variabl e`ZEPHYR_BUILD_COLOUR_DIAGNOSTIC` is set prior to starting a clean build. see #34254.

tejlmand commented:
@arvinf I see you added this:
```
if($ENV{ZEPHYR_BUILD_COLOUR_DIAGNOSTIC})
```
after approved but before merging.
I don't think we should introduce such environment flags without documenting them.
Secondly, why not a Kconfig setting ?

Not sure how many users will discover that they can actually enable colored output.

arvinf commented:
My reasoning was that it should be an environment variable because it affects the build process (similar to the choice of compiler) and isn't a feature that affects the firmware. 

It could be added as a default on to `zephyr-env.sh` assuming that doesn't affect the CI and build scripts.

I am OK with adding it to documentation. I just don't know the process for that. The best place for it would be in the doc for setting up the build system [https://docs.zephyrproject.org/latest/application/index.html]?

dcpleung commented:
I think this can be added to the Getting Started guide.

tejlmand commented:
> My reasoning was that it should be an environment variable rather than Kconfig because it affects the build process (similar to the choice of compiler) and isn't a feature that affects the firmware.

How does this affect build process in a different way that [Optimization](https://docs.zephyrproject.org/latest/reference/kconfig/choice_485.html), [CONFIG_COMPILER_OPT](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_COMPILER_OPT.html) or [CONFIG_CODING_GUIDELINE_CHECK](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_CODING_GUIDELINE_CHECK.html) ?

Even if not using Kconfig, then an environment variable should still not be used in this.
If there is a good reason, a Cmake cache variable that user can specify using `-D<var>=<val>` will still be better.
But from the looks of it, my preference here would be a Kconfig, which will give you generated documentation for free.

arvinf commented:
If it is a Kconfig variable then it would end up in `prj.conf` of the Application and the coloured output would affect CI and other automated builds for the end user (in their own environemnt). This is the issue raised by @dcpleung in his patch which prompted me to add the variable here. Am I mistaken in my understanding of this issue?

tejlmand commented:
> If it is a Kconfig variable then it would end up in prj.conf

Only if the user adds it there.
Users can do: `west build -b <board> -- -DCONFIG_COMPILER_COLOUR_DIAGNOSTIC=y`
and nothing is added to the `prj.conf`

This is a compiler setting and should be treated as any other such setting and not an environment variable.

arvinf commented:
I'll change it to Kconfig `CONFIG_BUILD_COLOUR_DIAGNOSTIC` and submit a new PR.

dcpleung commented:
This compiler option is to colorize compilation warnings and errors to the console, and does not affect the binary output. It makes the error messages easier to recognize so we don't have to mentally search through a sea of text with the same color. If you make this into kconfig, it means that every time I do west build, I need to manually add the config to the command line. Moreover, if I run twister locally, those tests which fail to build no longer have colorize build log as the kconfig is not enabled by default, and I have to manually add it. If we use an environment variable to enable colorized output, I can set this up in `~/.zephyrrc` and the output will be colorized for all builds (both west and twister) automatically. So I think having it to be enabled via environment variable is also important.

arvinf commented:
@dcpleung Are you propsing that we have both Kconfig and environment variable controls and that either could be set to enable this feature?

dcpleung commented:
Personally I think just using environment variable is enough, as this does not affect the binary output at all.

tejlmand commented:
> if you make this into kconfig, it means that every time I do west build, I need to manually add the config to the command line.

No, you can do:
```
west config build.cmake-args -- -DCONFIG_COMPILER_COLOUR_DIAGNOSTIC=y
```
and then it will be applied to CMake whenever you run `west build`.

And you can even make it global for all `west` workspaces by doing: `west config --global ....`

tejlmand commented:
> Personally I think just using environment variable is enough, as this does not affect the binary output at all.

This is also about having things consistent, and Kconfig / menuconfig provides a easy way for users to lookup all config options, without having to look into the docs.
And grouped with other compiler options, it will be grouped logically and show that color output can be enabled
I'm not sure why @arvinf decided not to just have it always on, as that was accepted here: https://github.com/zephyrproject-rtos/zephyr/pull/25557#issuecomment-633449791.

Also remember, environment settings are very common for Linux developers, but less common for windows developers.

There will always be someone who think a particular flag deserve to be an environment variable, but we should really keep those at a minimum, and so far I haven't seen a good enough reason why setting of a compile flag should be controlled through environment.

The toolchain is a very different story, because we need toolchain / compiler even before we start processing Kconfig.

arvinf commented:
@tejlmand How about we have it enabled by default (per original approved commit) but have a Kconfig to disable it? 

This way it is usable out of the box without the user needing to do anything or make changes to their `prj.conf`. For CI builds and other automated cases where colour output might interfere they can add `-DCONFIG_NO_COMPILER_COLOUR_DIAGNOSTIC=y` to their build scripts.

tejlmand commented:
> @tejlmand How about we have it enabled by default (per original approved commit) but have a Kconfig to disable it?

I think that sounds very reasonable, and will most likely suit majority of users if not all users.

arvinf commented:
Done. See #34498
